Generate Package Open Graph Images for Social Links

TammyTee opened the pull request:
## Summary
This PR: 
* adds a new job that is responsible for generating images for packages that can be used in their open graph meta. The job is dispatched when a call to the `packages.show` route is made.
* adds in a base image that the open graph images are created from

**Question:** The images are currently being saved to storage and a symlink was created. Are there downsides to this as opposed to saving directly to `public/`?

## Screenshots
![image](https://user-images.githubusercontent.com/8689444/79246532-501bd300-7e47-11ea-8e7d-42a569285977.png)

![image](https://user-images.githubusercontent.com/8689444/79249211-f0272b80-7e4a-11ea-8401-77c9f5a07414.png)


### Testing using [MetaTags.io](https://metatags.io/)

***Note: the Slack screenshot is inaccurate, looks good in the actual app***

![metatags io_](https://user-images.githubusercontent.com/8689444/79246883-c3bde000-7e47-11ea-8e0b-3a2e9473ec3e.png)

mattstauffer commented:
> Question: The images are currently being saved to storage and a symlink was created. Are there downsides to this as opposed to saving directly to public/?

Nope, this is the better way!

mattstauffer commented:
Things I want to consider without having read the code:
- How does it handle longer package titles?
- If they rename the package title, does this get regenerated?

Excited to read the code!

mattstauffer commented on app/Jobs/GeneratePackageOpenGraphImage.php:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @return void
+     */
+    public function __construct(String $packageName, String $packageAuthor)
```
How hard would it be to pass an instance of the Package model instead of the individual strings here? The one nice thing about that is that then we could add the open graph image name code (with the Str::slug) as a method on the package.

mattstauffer replied:
_or_, its third param could be the OG name, and then this could become even dumber

mattstauffer commented on app/Jobs/GeneratePackageOpenGraphImage.php:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @return void
+     */
+    public function handle()
```
I'd pull each of the responsibilities of this job at *least* into its own method here. Maybe even separate classes, but let's start with their own methods.

So, what does this method do?

1. Ensures a directory exists
2. Checks if the file already exists
3. Does a crap ton of image generate-y stuff
4. Saves the file

I'd love to have at *least* a separate method (maybe a class but I'm not sure) which has three params: title, name, and og path. This method generates the image and saves it to that path, without any of the other logic in here.

mattstauffer commented on tests/Feature/GeneratePackageOpenGraphImageJobTest.php:
```diff
@@ -0,0 +1,59 @@
+use Illuminate\Support\Facades\Storage;
+use Tests\TestCase;
+
+/** @runTestsInSeparateProcesses */
```
Uh TIL

TammyTee commented:
@mattstauffer here's a break down of what changed:

- created commands for purging and generating the images (can optionally pass the package ID to each command)
`php artisan purge:ogimage`
`php artisan generate:ogimage`

- extracted image creation functionality from the job into a separate class
- created a config file for the open graph settings (specifically the directory name where they'll be stored so that there isn't a "magic string" floating across the app)
- ensured that longer package titles are accounted for in all cases. Here's a worst case example:

![image](https://user-images.githubusercontent.com/8689444/80988048-f03e9980-8e00-11ea-91c2-e861d3419676.png)

- changed the job to be dispatched when the package is created AND updated so that we can catch any changes to the name etc.

damiani commented on app/OpenGraphImage.php:
```diff
@@ -0,0 +1,125 @@
+        return true;
+    }
+
+    public function removeDuplicates()
```
"duplicates" is a little misleading in this method name, since neither the filenames are duplicate, nor the image contents (if the package name is different). Maybe something like `deleteImagesForProject`? That's super cumbersome, but more accurate. What do you think?

TammyTee replied:
Agree. Thank you! I struggled naming this one.

damiani commented on app/OpenGraphImage.php:
```diff
@@ -0,0 +1,125 @@
+    public function save($image)
+    {
+        imagepng($image, $this->storagePath($this->fileName));
+
```
```suggestion
```

TammyTee commented on app/Console/Commands/GenerateOpenGraphImages.php:
```diff
@@ -0,0 +1,43 @@
+            : $this->callSilent('purge:ogimage');
+
+        $packages = $this->argument('package')
+            ? Package::where('id', $this->argument('package'))->get()
```
Changing this to `::find` as opposed to `::where(...)->get()` (which returns a collection) causes the `count` happening on line 25 to fail and looping over the `$packages` on line 31 no longer works either ... error of "Trying to get properties of a non-object"

damiani replied:
Ah right, I didn't notice it was `get()` ... all good, then :)

TammyTee commented on app/OpenGraphImage.php:
```diff
@@ -0,0 +1,125 @@
+    public static function makeFileName($uniqueId, $name): string
+    {
+        return "{$uniqueId}_" . Str::slug($name, '-') . '.png';
+    }
```
Tlint complains about this since the method is static? Should I swap and ignore Tlint?

TammyTee replied:
But I guess this doesn't really matter if I create the `generate()` method (that you suggested above) thats being called statically ... I could get rid of the constructor all together.

damiani replied:
So, TLint is wrong here, apparently :)

However, rather than use a static method here, let's instead keep the constructor, accept all 3 parameters, and add a non-static `generate` method:

```
(new OpenGraphImage($this->packageName, $this->packageAuthor, $this->packageOgImageName))->generate();
```

Every other method in this class, then, can be `protected`, since they're only used internally to the class.

damiani commented on app/OpenGraphImage.php:
```diff
@@ -25,32 +44,38 @@ public static function makeFileName($uniqueId, $name): string
+
+        if (! $this->makeStorageDirectory()) {
+            $this->deleteImagesForProject();
+        }
```
I think we can get rid of `makeStorageDirectory()` (see comment below), and just always check for other project images. So the contents of this could become:


```php
Storage::ensureDirectoryExists($this->storageDirectory());
$this->deleteImagesForProject();
$this->save($this->make());
```

damiani commented on app/OpenGraphImage.php:
```diff
@@ -25,32 +44,38 @@ public static function makeFileName($uniqueId, $name): string
         }
 
-        Storage::makeDirectory($this->storageDirectory());
+        Storage::ensureDirectoryExists($this->storageDirectory());
```
This call would actually replace the whole `makeStorageDirectory()` function; it checks if it exists, and creates it if it doesn't.

TammyTee commented:
Hey @damiani this should be all good here now 🤞. 
And it actually looks like the `ensureDirectoryExists` method lives on the `File` facade so I had to tweak that a bit.
